src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -331,6 +331,13 @@ private void HandleBeforeHandshakeCompletesException(IConnection connection, Exc
                     return; // stale generation number
                 }
 
+                if (ex is MongoConnectionException mongoConnectionException &&
+                    mongoConnectionException.IsNetworkException &&
+                    !mongoConnectionException.ContainsSocketTimeoutException)
```

Comment:
Can a connection exception also be a socket time exception?  And if so: if connect times out, would we not still want to cancel the current check?

Reply:
>Can a connection exception also be a socket time exception?

If you ask about how it's implemented in the C#, then yes. MongoConnectionException  contains different cases because actually this exception wraps a number of cases from BinaryConnection https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L725-L744

As far as I know, ContainsSocketTimeoutException is the connection timeout exception with the code: `SocketError.TimedOut`. The spec says that we need to cancel only `Network error`.

This is a note: `When an application operation fails with a non-timeout network error, drivers cancel that monitor's in-progress check.` from here:https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-monitoring.rst#drivers-cancel-in-progress-monitor-checks

It also mimics the same logic in AFTER handshake error processing. Do you think we need to reconsider this logic?

Reply:
I don't really know.  I will defer to you and @rstam

Reply:
@dima what exception is thrown if a connection times out? Do we need to tell the difference between a timeout in the initial connection attempt vs a timeout in a subsequent I/O operation (like isMaster)? If we need to tell the difference can we?

Reply:
In two words, I would say, no changes are required right now, but we need to investigate why the TimeoutException from the driver side was not considered for this logic (I think it will also require changes here: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Servers/Server.cs#L301)

Reply:
Vincent's comment about ContainsSocketTimeoutException: https://github.com/vincentkam/mongo-csharp-driver/pull/114#discussion_r428173287
So, I think both of the above points should be considered as Timeout exception. But I still think it's better to change it in the scope of a new ticket